File: ioi2020-biscuits/solution/biscuits-yanhao-ac_bottom_up.java
class biscuits {	
	long[] dp = new long[61];
	long[] b = new long[61];
	
	long count_tastiness(long x, long[] a) {
		long sum = 0;
		int last = Long.numberOfLeadingZeros(x)-3;
		for(int i=0; i<=last; i++) {
			dp[i] = 0;
			long j = (1L<<i);
			for(int h=i-1; h>=0; h--) {
				if((j>>h) > 0) {
					dp[i] += dp[h];
					j = (j < b[h] ? j : b[h])-(1L<<h);
				}
			}
			if(j==1) dp[i]++;
			sum += (i<a.length ? (a[i]<<i) : 0);
			b[i] = sum/x+1;
		}
		return dp[last];
	}
}

File: ioi2020-biscuits/solution/biscuits-yanhao-ac_bottom_up.cpp
#include "biscuits.h"
#include <vector>

long long dp[61];
long long b[61];
long long count_tastiness(long long x, std::vector<long long> a) {
    long long sum = 0;
    for(int i=0; i<=60-std::__lg(x); i++) {
        dp[i] = 0;
        long long j = (1LL<<i);
        for(int h=i-1; h>=0; h--) {
            if((j>>h)>0) {
                dp[i] += dp[h];
                j = std::min(j, b[h])-(1LL<<h);
            }
        }
        dp[i] += (j==1);
        sum += (i<(int)a.size() ? (a[i]<<i) : 0);
        b[i] = sum/x+1;
    }
    return dp[60-std::__lg(x)];
}
